File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
def solution(genres, plays):
    answer = []
    summary_dict = {}
    # {{'gen': 'pop', 'total': 170, 'list': [(-2500,2), (-2500,4)]},
    # {'gen': 'pop', 'total': 170, 'list': [(-2500,2), (-2500,4)]},}
    
    for idx, (genre, play) in enumerate(zip(genres, plays)):
        if genre not in summary_dict:
            summary_dict[genre]= {
                'total': 0, 
                'list': []
            }
        summary_dict[genre]['total']+=play
        summary_dict[genre]['list'].append((-play, idx))
    
    summary_dict = sorted(summary_dict.items(), key = lambda x: x[1]['total'], reverse=True)
    for gen, info in summary_dict:
        info['list'].sort(key=lambda x: x[0])
    
    for gen, info in summary_dict:
        answer.append(info['list'][0][1])
        if len(info['list'])>1:
            answer.append(info['list'][1][1])
        
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;
    unordered_map<string, vector<vector<int>>> playlist;
    vector<pair<int, string>> order;
    
    //playlist 해시맵 만들기
    for(int i=0; i<genres.size(); i++){
        playlist[genres[i]].push_back({plays[i], -i});
    }
    
    for(auto [name, play]: playlist){
        int total=0;
        for (auto p : play){
            total+=p[0];
        }
        order.push_back({total, name});
    }
    
    sort(order.rbegin(), order.rend());
    
    for(auto& p:playlist){
        sort(p.second.rbegin(), p.second.rend());
    }
    
    for(auto [total, name]: order){
        int cnt = 0;
        for(auto song: playlist[name]){
            answer.push_back(-song[1]);
            cnt++;
            if (cnt==2){
                break;
            }
        }
        
    }
    
    return answer;
}